File: algorithms/2_medium/M/02988_minimum_number_of_operations_to_make_x_and_y_equal/dfs_memo.c
static inline int min_int(const int a, const int b) {
    return a < b ? a : b;
}

static int dfs(const int x, const int y, int* memo) {
    if (x <= y) {
        return y - x;
    }

    if (memo[x] != -1) {
        return memo[x];
    }

    int res = x - y;
    if (x % 5 == 0) {
        res = min_int(res, 1 + dfs(x / 5, y, memo));
    } else {
        const int diff = x % 5;
        res = min_int(res, 1 + diff + dfs(x / 5, y, memo));
        res = min_int(res, 1 + 5 - diff + dfs(1 + (x / 5), y, memo));
    }

    if (x % 11 == 0) {
        res = min_int(res, 1 + dfs(x / 11, y, memo));
    } else {
        const int diff = x % 11;
        res = min_int(res, 1 + diff + dfs(x / 11, y, memo));
        res = min_int(res, 1 + 11 - diff + dfs(1 + (x / 11), y, memo));
    }

    memo[x] = res;
    return res;
}

int minimumOperationsToMakeEqual(int x, int y) {
    if (x > 10000 && x < 1 && y < 1 && y > 10000) {
        fprintf(stderr, "invalid input");
        exit(1);
    }

    if (x <= y) {
        return y - x;
    }

    int* memo = malloc((size_t)(x + 1) * sizeof(*memo));
    if (memo == NULL) {
        fprintf(stderr, "alloc fail");
        exit(1);
    }
    for (int i = 0; i <= x; i += 1) { memo[i] = -1; }

    const int res = dfs(x, y, memo);
/* cleanup: */
    free(memo);
    return res;
}

File: algorithms/2_medium/M/02988_minimum_number_of_operations_to_make_x_and_y_equal/dfs_memo.rs
impl Solution {
    fn dfs(x: i32, y: i32, memo: &mut[i32]) -> i32 {
        if x <= y {
            return y - x;
        }

        if memo[x as usize] != -1 {
            return memo[x as usize];
        }

        let mut res = x - y;
        if x % 5 == 0 {
            res = res.min(1 + Self::dfs(x / 5, y, memo));
        } else {
            let diff = x % 5;
            res = res.min(1 + diff + Self::dfs(x / 5, y, memo))
                     .min(1 + 5 - diff + Self::dfs(1 + (x / 5), y, memo));
        }

        if x % 11 == 0 {
            res = res.min(1 + Self::dfs(x / 11, y, memo));
        } else {
            let diff = x % 11;
            res = res.min(1 + diff + Self::dfs(x / 11, y, memo))
                     .min(1 + 11 - diff + Self::dfs(1 + (x / 11), y, memo));
        }

        memo[x as usize] = res;
        res
    }

    pub fn minimum_operations_to_make_equal(x: i32, y: i32) -> i32 {
        assert!(x <= 10_000 && x >= 1 && y >= 1 && y <= 10_000);
        if x <= y {
            return y - x;
        }

        let mut memo = vec![-1_i32; x as usize + 1];

        Self::dfs(x, y, &mut memo)
    }
}
